src/Compiler/Service/ServiceParseTreeWalk.fs

```diff
@@ -575,6 +582,14 @@ module SyntaxTraversal =
 
                     if ok.IsSome then ok else traverseSynExpr synExpr
 
+                | SynExpr.Lambda (inLambdaSeq = false; body = synExpr; parsedData = Some (pats, _)) ->
```

Comment:
This feels weird to introduce without a code comment.
Why is the `bodyExpr` from `parsedData` not used?
The [inner lambda](https://fsprojects.github.io/fantomas-tools/#/ast?data=N4KABGBEAmCmBmBLAdrAzpAXFSAacUiaAYmolmPAIYA2as%2BEkAxgPZwWTwCuyYAHmACeYAF5gAtAD4wACgCUeApFj8ADlWTQKAFwBO3WCAC%2BQA) will be processed again, so I believe the pats will be visited multiple times here.

Reply:
…But as for why I didn't add traversal of the `bodyExpr` of `parsedData`: it seems like it's the same as the outer `body` (both are just `SynExpr`s). Do you know whether there is any case where it isn't?

Reply:
See https://fsprojects.github.io/fantomas-tools/#/ast?data=N4KABGBEAmCmBmBLAdrAzpAXFSAacUiaAYmolmPAIYA2as%2BEkAxgPZwWTwCuyYAHmACeYAF5gAtAD4wACgCUeApFj8ADlWTQKAFwBO3WCAC%2BQA

```fsharp
fun x y z -> ()
```

![image](https://github.com/dotnet/fsharp/assets/2621499/d3bb4e3c-76e9-4ce4-ab68-d858228da227)

Reply:
Ah, yes. Does f7403f4 look better? I don't think we need to traverse `bodyExpr` since it will eventually be visited as part of `body`.

Reply:
Hmm, I'm not overly sure but I think in general it might make more sense to just traverse `bodyExpr` and not bother with `body` in this case.

This question is larger than the scope of this PR. What was already in the visitor code was questionable to begin with.

The untyped AST for Lambdas is already optimized to work with during type-checking.

When a user has `fun x y z -> ()`, the AST is already expanded:

```fsharp
SynModuleDecl.Expr(
    expr =
        SynExpr.Lambda(
            body =
                SynExpr.Lambda(
                    body =
                        SynExpr.Lambda(
                            body = SynExpr.Const(constant = SynConst.Unit, range = R("(1,13--1,15)")),
                        ),
                ),
            parsedData =
                Some(
                    ([
                        SynPat.Named(
                            ident = SynIdent.SynIdent(ident = Ident("x", R("(1,4--1,5)")), trivia = None),
                            isThisVal = false,
                            accessibility = None,
                            range = R("(1,4--1,5)")
                        )
                        SynPat.Named(
                            ident = SynIdent.SynIdent(ident = Ident("y", R("(1,6--1,7)")), trivia = None),
                            isThisVal = false,
                            accessibility = None,
                            range = R("(1,6--1,7)")
                        )
                        SynPat.Named(
                            ident = SynIdent.SynIdent(ident = Ident("z", R("(1,8--1,9)")), trivia = None),
                            isThisVal = false,
                            accessibility = None,
                            range = R("(1,8--1,9)")
                        )
                     ],
                     SynExpr.Const(constant = SynConst.Unit, range = R("(1,13--1,15)")))
                ),
            range = R("(1,0--1,15)"),
            trivia = { ArrowRange = Some(R("(1,10--1,12)")) }
        ),
    range = R("(1,0--1,15)")
)
```

The user did not write `fun x -> fun y -> fun z`, so I wouldn't follow that path to begin with in the visitor code.

I cannot think of any scenarios where you would want to do this from a tooling perspective. Although maybe I lack imagination.

I'm not hellbent on this, this is just my take on it.
@baronfel what are your thoughts on this one?

Reply:
8e493e1